influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -108,59 +125,213 @@ struct CacheValue {
     }
 }
 
 /// The state of a cache entry
 #[derive(Debug)]
-enum CacheEntry {
+enum CacheEntryState {
     /// The cache entry is being fetched from object store
-    Fetching,
+    ///
+    /// This holds a [`watch::Sender`] that is used to notify requests made to get this entry
+    /// while it is being fetched by the cache oracle.
+    Fetching(watch::Sender<Option<Arc<CacheValue>>>),
     /// The cache entry was successfully fetched and is stored in the cache as a [`CacheValue`]
     Success(Arc<CacheValue>),
-    /// The request to the object store failed
-    Failed,
-    /// The cache entry was deleted
-    Deleted,
-    /// The object is too large for the cache
-    TooLarge,
 }
 
-impl CacheEntry {
-    /// Get the size of thje cache entry in bytes
+impl CacheEntryState {
+    /// Get the approximate size of the cache entry in bytes
     fn size(&self) -> usize {
         match self {
-            CacheEntry::Fetching => 0,
-            CacheEntry::Success(v) => v.size(),
-            CacheEntry::Failed => 0,
-            CacheEntry::Deleted => 0,
-            CacheEntry::TooLarge => 0,
+            CacheEntryState::Fetching(tx) => std::mem::size_of_val(tx),
+            CacheEntryState::Success(v) => v.size(),
         }
     }
+}
 
-    fn is_fetching(&self) -> bool {
-        matches!(self, CacheEntry::Fetching)
+/// A cache for storing objects from object storage by their [`Path`]
+///
+/// This acts as a Least-Frequently-Used (LFU) cache that allows for concurrent reads. See the
+/// [`Cache::prune`] method for implementation of how the cache entries are pruned. Pruning must
+/// be invoked externally, e.g., on an interval.
+#[derive(Debug)]
+struct Cache {
+    /// The maximum amount of memory this cache should occupy
+    capacity: usize,
+    /// The current amount of memory being used by the cache
+    used: AtomicUsize,
+    /// What percentage of the total number ofo cache entries will be pruned during a pruning operation
+    prune_percent: f64,
+    /// The map storing cache entries
+    map: DashMap<Path, CacheEntry>,
+    time_provider: Arc<dyn TimeProvider>,
+}
+
+impl Cache {
+    /// Create a new cache with a given capacity and max free percentage
+    ///
+    /// The cache's `max_free_amount` will be calculated using the provided values
+    fn new(capacity: usize, prune_percent: f64, time_provider: Arc<dyn TimeProvider>) -> Self {
+        Self {
+            capacity,
+            used: AtomicUsize::new(0),
+            prune_percent,
+            map: DashMap::new(),
+            time_provider,
+        }
     }
 
-    fn is_success(&self) -> bool {
-        matches!(self, CacheEntry::Success(_))
+    /// Get an entry in the cache or `None` if there is not an entry
+    ///
+    /// If the entry is in `Fetching` state, then this will await the etry having been fetched.
+    async fn get(&self, path: &Path) -> Option<Arc<CacheValue>> {
+        let entry = self.map.get(path)?;
+        match &entry.state {
```

Comment:
Could we handle "fetching vs fetched" outside the cache maybe? In that way, cache either has something or not. It's up to the call site (object store impl) to decide if there are multiple requests to the same path (and that could call them internally however it pleases by holding semaphore or whatever) and then finally it just loads the data into the cache as a single op.

Reply:
I will try this out - I guess the idea is to have the signature:
```rust
fn get(&self, path: &Path) -> Option<CacheEntryState> {
    /* ... */
}
```
Then the caller decides what to do with the state. Perhaps I can tie that in with Paul's above feedback on the current use of `watch::channel`.

Reply:
I think I accomplished this in https://github.com/influxdata/influxdb/pull/25389/commits/667eab588b80adf8049a5bec8cf43a0612d47560 - the `get` method now just returns an `Option<CacheEntryState>`, which takes a clone of the `CacheEntryState`, which should be cheap, and returns that if the entry exists. This should free up the reference to the map immediately, and makes `get` non-async, so the caller needs to wait on the `Fetching` future if it is in that state.